pageserver/src/tenant/storage_layer/image_layer.rs

```diff
@@ -534,6 +558,52 @@ impl ImageLayerInner {
         Ok(planner.finish())
     }
 
+    /// Given a key range and filter function, generate a read plan and execute it, then for
+    /// each value we read write it onwards into the writer.
+    pub(super) async fn do_reads_and_write<F>(
```

Comment:
A test for this would be nice. You could look the tests in `delta_layer.rs` for the harness.